### DISCUSSION TITLE
Tips on try/catch error.

### DISCUSSION THREAD

**Discussion by alucardu** [prisma/prisma#18009] [Q&A]
```javascript
export const userResolvers = {
  Mutation: {
    createUser: async (_, args) => {
      const sessionUsers = await prisma.session.findUnique({
        where: {
          id: Number(args.sessionId)
        },
        include: {
          players: true
        }
      })

      try {
        if (sessionUsers.players.some((user) => user.name === args.name)) {
          throw 'Error'
        }

        return await prisma.user.create({
          data: {
            name: args.name,
            session: {
              connect: { id: Number(args.sessionId) },
            }
          }
        })
      } catch (err) {
        throw new UserInputError(10004);
      }
    },
  },
};
```
In my app multiple sessions can be created, and each session can have many users. But I don't want multiple users with the same user name in the session. So in the create user resolver I check if the passed in user name already exists if it does I throw `Error`, if not I create the user.

It's working, but I do wonder if this is the correct way of building a exception handler.

**Comment by nurul3101** [comment]
Hello @alucardu 👋 

One suggestion I have is that Instead of throwing a generic string error, consider using a specific error type like `Error`. Using a specific error type can make it easier to handle errors in your code and provide better error messages to the client.

You can do something like this:
```typescript
if (sessionUsers.players.some((user) => user.name === args.name)) {
        throw new Error('Username already exists');
}
```

To simplify this, I would also recommend that you can consider adding a `@unique` attribute to your name field in User model. This way database would throw an error if a user with same name exists.

**Reply by alucardu** [reply]
Thank you for the suggestion. The reason I'm sending a number instead of a string is because I want to show the error in the front end based on a number. Also the username field is not unique. It should be unique to the session.

**Reply by nurul3101** [reply]
Then I think adding a unique index on combination of session_id and username could help.

**Reply by alucardu** [reply]
That sounds like a great tip, I will investigate.

**Reply by nurul3101** [reply]
Here's the [guide](https://www.prisma.io/docs/concepts/components/prisma-schema/data-model#defining-a-unique-field) on defining unique index on multiple columns in prisma schema. I believe this should be helpful.

**Comment by alucardu** [comment]
The unique field tip was what I needed,

```
prisma.schema

model User {
  id            Int         @id       @default(autoincrement())
  name          String
  session       Session?    @relation(fields: [sessionId], references: [id])
  sessionId     Int?
  @@unique([sessionId, name])
}

model Session {
  id            Int         @id       @default(autoincrement())
  name          String
  code          String      @unique
  players       User[]
}
```

Now in my resolver I don't have to manually look for an existing user:

```javascript
export const userResolvers = {
  Mutation: {
    createUser: async (_, args) => {
      try {
        return await prisma.user.create({
          data: {
            name: args.name,
            session: {
              connect: { id: Number(args.sessionId) },
            }
          }
        })
      } catch (err) {
        throw new Error('Username already exists');
      }
    },
  },
};
```

**Reply by nurul3101** [reply]
I am glad to hear that combined unique index worked for your use case! 🚀

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
